File: Data Structures & Algorithms/validate-parentheses/submission-2.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        for c in s:
            if c == '(' or c == '[' or c == '{':
                st.append(c)
            else:
                if not st:
                    return False
                top = st[-1]
                if c == ')' and top != '(':
                    return False
                if c == ']' and top != '[':
                    return False
                if c == '}' and top != '{':
                    return False
                st.pop()        

        return not st

File: Data Structures & Algorithms/task-scheduling/submission-2.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        // priority_queue<pair<int, char>>, vector<pair<int, char>> pq; // maxHeap
        // select the maximum amount task
        // queue<tuple<int, char, int>> cooldown;
        // int timestamp;

        unordered_map<char, int> freq;

        for(auto t : tasks) freq[t]++;

        priority_queue<pair<int, char>, vector<pair<int, char>>>  pq; // maxHeap
        // select the maximum amount task

        queue<tuple<int, char, int>> cooldown; // cooldown queue

        for (auto [t, f] : freq) pq.push({f, t});

        int timestamp = 0;
        while (!pq.empty() || !cooldown.empty()) {
            // any task can be add back to pq or not
            if (!cooldown.empty()) {
                auto [f, t, returnTime] = cooldown.front();
                if (returnTime < timestamp) {
                    pq.push({f, t});
                    cooldown.pop();
                }
            }

            // pick the maximum amount task
            if (!pq.empty()) {
                auto [f, t] = pq.top(); pq.pop();
                // cout << timestamp << " " << t << endl;
                if (f > 1) {
                    cooldown.push({f - 1, t, timestamp + n});
                }
            }
            timestamp++;
        }

        return timestamp;
    }
};
